[rosidl_parser] Some IDL attributes cause rosidl_parser to fail

<!--
For general questions, please ask on ROS answers: https://answers.ros.org, make sure to include at least the `ros2` tag and the rosdistro version you are running, e.g. `ardent`.
For general design discussions, please post on discourse: https://discourse.ros.org/c/ng-ros
Not sure if this is the right repository? Open an issue on https://github.com/ros2/ros2/issues
For Bug report or feature requests, please fill out the relevant category below
-->

## Bug report

**Required Info:**

- Operating System:
  - Ubuntu 16.04
- Installation type:
  - Source
- Version or commit hash:
  - ROS2 Dashing
- DDS implementation:
  - RTI Connext
- Client library (if applicable):
  - <!-- e.g. rclcpp, rclpy, or N/A -->

#### Steps to reproduce issue
<!-- Detailed instructions on how to reliably reproduce this issue http://sscce.org/
``` code that can be copy-pasted is preferred ``` -->
Use the following IDL and try to use it in a message package. It will fail before it gets to the DDS code generator, so you can reproduce this even if you do not have a DDS implementation that supports this particular annotation
```
module testmsgs {
  module msg {
    @transfer_mode(SHMEM_REF)
    struct IdlMessage {
      @key
      int32 int_value;
    };
  };
};
```

#### Expected behavior
Building a package that contains this message definition should succeed

#### Actual behavior
```
--- stderr: std_msgs                            
Tree(const_expr, [Tree(or_expr, [Tree(xor_expr, [Tree(and_expr, [Tree(shift_expr, [Tree(add_expr, [Tree(mult_expr, [Tree(unary_expr, [Tree(primary_expr, [Tree(scoped_name, [Token(ID
ENTIFIER, 'SHMEM_REF')])])])])])])])])])]) /home/kyle.marcey/ros2/src/ros2/common_interfaces/std_msgs/msg/IdlMessage.idl
Error processing idl file: /home/kyle.marcey/ros2/src/ros2/common_interfaces/std_msgs/msg/IdlMessage.idl
Traceback (most recent call last):
  File "/home/kyle.marcey/ros2/install/rosidl_generator_c/share/rosidl_generator_c/cmake/../../../lib/rosidl_generator_c/rosidl_generator_c", line 40, in <module>
    sys.exit(main())
  File "/home/kyle.marcey/ros2/install/rosidl_generator_c/share/rosidl_generator_c/cmake/../../../lib/rosidl_generator_c/rosidl_generator_c", line 35, in main
    args.generator_arguments_file,
  File "/home/kyle.marcey/ros2/install/rosidl_generator_c/lib/python3.6/site-packages/rosidl_generator_c/__init__.py", line 37, in generate_c
    generate_files(generator_arguments_file, mapping)
  File "/home/kyle.marcey/ros2/install/rosidl_cmake/lib/python3.6/site-packages/rosidl_cmake/__init__.py", line 90, in generate_files
    raise(e)
  File "/home/kyle.marcey/ros2/install/rosidl_cmake/lib/python3.6/site-packages/rosidl_cmake/__init__.py", line 70, in generate_files
    idl_file = parse_idl_file(locator)
  File "/home/kyle.marcey/ros2/install/rosidl_parser/lib/python3.6/site-packages/rosidl_parser/parser.py", line 64, in parse_idl_file
    content = parse_idl_string(string, png_file=png_file)
  File "/home/kyle.marcey/ros2/install/rosidl_parser/lib/python3.6/site-packages/rosidl_parser/parser.py", line 73, in parse_idl_string
    content = extract_content_from_ast(tree)
  File "/home/kyle.marcey/ros2/install/rosidl_parser/lib/python3.6/site-packages/rosidl_parser/parser.py", line 142, in extract_content_from_ast
    annotations = get_annotations(struct_defs[0])
  File "/home/kyle.marcey/ros2/install/rosidl_parser/lib/python3.6/site-packages/rosidl_parser/parser.py", line 517, in get_annotations
    value = get_const_expr_value(const_expr)
  File "/home/kyle.marcey/ros2/install/rosidl_parser/lib/python3.6/site-packages/rosidl_parser/parser.py", line 527, in get_const_expr_value
    assert len(literals) == 1, str(const_expr)
AssertionError: Tree(const_expr, [Tree(or_expr, [Tree(xor_expr, [Tree(and_expr, [Tree(shift_expr, [Tree(add_expr, [Tree(mult_expr, [Tree(unary_expr, [Tree(primary_expr, [Tree(scoped
_name, [Token(IDENTIFIER, 'SHMEM_REF')])])])])])])])])])])
make[2]: *** [rosidl_generator_c/std_msgs/msg/idl_message.h] Error 1
```

#### Additional information

I believe the problem is that the function `get_const_expr_value` in `rosidl_parser/parser.py` only handles `literals`, but the grammar allows either a `literal` or a `scoped_name`

--- a/rosidl_parser/rosidl_parser/parser.py
+++ b/rosidl_parser/rosidl_parser/parser.py
@@ -522,7 +522,6 @@ def get_annotations(tree):
 
 
 def get_const_expr_value(const_expr):
-    # the const_expr could either be a literal or a name
     # TODO support arbitrary expressions
     expr = list(const_expr.find_data('primary_expr'))
     assert len(expr) == 1, str(expr)
@@ -573,7 +572,7 @@ def get_const_expr_value(const_expr):
 
         assert False, 'Unsupported tree: ' + str(const_expr)
     else:
-        assert False, 'Unknown type: ' + str(child.data)
+        assert False, 'Unsupported tree: ' + str(const_expr)
 
 
 def get_decimal_literal_value(decimal_literal):

--- a/rosidl_parser/rosidl_parser/parser.py
+++ b/rosidl_parser/rosidl_parser/parser.py
@@ -528,9 +528,9 @@ def get_const_expr_value(const_expr):
     primary_expr = expr[0]
     assert len(primary_expr.children) == 1
     child = primary_expr.children[0]
-    if child.data == 'scoped_name':
+    if 'scoped_name' == child.data:
         return str(child.children[0])
-    elif child.data == 'literal':
+    elif 'literal' == child.data:
         literals = list(const_expr.find_data('literal'))
         # TODO support arbitrary expressions
         assert len(literals) == 1, str(const_expr)

--- a/rosidl_parser/rosidl_parser/parser.py
+++ b/rosidl_parser/rosidl_parser/parser.py
@@ -523,6 +523,7 @@ def get_annotations(tree):
 
 def get_const_expr_value(const_expr):
     # the const_expr could either be a literal or a name
+    # TODO support arbitrary expressions
     expr = list(const_expr.find_data('primary_expr'))
     assert len(expr) == 1, str(expr)
     primary_expr = expr[0]
@@ -531,15 +532,12 @@ def get_const_expr_value(const_expr):
     if 'scoped_name' == child.data:
         return str(child.children[0])
     elif 'literal' == child.data:
-        literals = list(const_expr.find_data('literal'))
-        # TODO support arbitrary expressions
-        assert len(literals) == 1, str(const_expr)
-
+        literal = child
         unary_operator_minuses = list(const_expr.find_data('unary_operator_minus'))
         negate_value = len(unary_operator_minuses) % 2
 
-        assert len(literals[0].children) == 1
-        child = literals[0].children[0]
+        assert len(literal.children) == 1
+        child = literal.children[0]
 
         if child.data == 'integer_literal':
             assert len(child.children) == 1

--- a/rosidl_parser/rosidl_parser/parser.py
+++ b/rosidl_parser/rosidl_parser/parser.py
@@ -522,49 +522,58 @@ def get_annotations(tree):
 
 
 def get_const_expr_value(const_expr):
-    literals = list(const_expr.find_data('literal'))
-    # TODO support arbitrary expressions
-    assert len(literals) == 1, str(const_expr)
+    # the const_expr could either be a literal or a name
+    expr = list(const_expr.find_data('primary_expr'))
+    assert len(expr) == 1, str(expr)
+    child = expr[0].children[0]
+    if child.data == 'scoped_name':
+        return str(child.children[0])
+    elif child.data == 'literal':
+        literals = list(const_expr.find_data('literal'))
+        # TODO support arbitrary expressions
+        assert len(literals) == 1, str(const_expr)
 
-    unary_operator_minuses = list(const_expr.find_data('unary_operator_minus'))
-    negate_value = len(unary_operator_minuses) % 2
+        unary_operator_minuses = list(const_expr.find_data('unary_operator_minus'))
+        negate_value = len(unary_operator_minuses) % 2
 
-    assert len(literals[0].children) == 1
-    child = literals[0].children[0]
+        assert len(literals[0].children) == 1
+        child = literals[0].children[0]
 
-    if child.data == 'integer_literal':
-        assert len(child.children) == 1
-        child = child.children[0]
+        if child.data == 'integer_literal':
+            assert len(child.children) == 1
+            child = child.children[0]
 
-        if child.data == 'decimal_literal':
-            value = get_decimal_literal_value(child)
+            if child.data == 'decimal_literal':
+                value = get_decimal_literal_value(child)
+                if negate_value:
+                    value = -value
+                return value
+
+            assert False, 'Unsupported tree: ' + str(child)
+
+        if child.data == 'floating_pt_literal':
+            value = get_floating_pt_literal_value(child)
             if negate_value:
                 value = -value
             return value
 
-        assert False, 'Unsupported tree: ' + str(child)
+        if child.data == 'boolean_literal':
+            assert len(child.children) == 1
+            child = child.children[0]
+            assert child.data in ('boolean_literal_true', 'boolean_literal_false')
+            return child.data == 'boolean_literal_true'
 
-    if child.data == 'floating_pt_literal':
-        value = get_floating_pt_literal_value(child)
-        if negate_value:
-            value = -value
-        return value
+        if child.data == 'string_literals':
+            assert not negate_value
+            return get_string_literals_value(child, allow_unicode=False)
 
-    if child.data == 'boolean_literal':
-        assert len(child.children) == 1
-        child = child.children[0]
-        assert child.data in ('boolean_literal_true', 'boolean_literal_false')
-        return child.data == 'boolean_literal_true'
+        if child.data == 'wide_string_literals':
+            assert not negate_value
+            return get_string_literals_value(child, allow_unicode=True)
 
-    if child.data == 'string_literals':
-        assert not negate_value
-        return get_string_literals_value(child, allow_unicode=False)
-
-    if child.data == 'wide_string_literals':
-        assert not negate_value
-        return get_string_literals_value(child, allow_unicode=True)
-
-    assert False, 'Unsupported tree: ' + str(const_expr)
+        assert False, 'Unsupported tree: ' + str(const_expr)
+    else:
+        assert False, 'Unknown type: ' + str(child.data)
 
 
 def get_decimal_literal_value(decimal_literal):

--- a/rosidl_parser/rosidl_parser/parser.py
+++ b/rosidl_parser/rosidl_parser/parser.py
@@ -525,7 +525,9 @@ def get_const_expr_value(const_expr):
     # the const_expr could either be a literal or a name
     expr = list(const_expr.find_data('primary_expr'))
     assert len(expr) == 1, str(expr)
-    child = expr[0].children[0]
+    primary_expr = expr[0]
+    assert len(primary_expr.children) == 1
+    child = primary_expr.children[0]
     if child.data == 'scoped_name':
         return str(child.children[0])
     elif child.data == 'literal':

--- a/rosidl_parser/test/msg/MyMessage.idl
+++ b/rosidl_parser/test/msg/MyMessage.idl
@@ -13,6 +13,7 @@ module rosidl_parser {
     };
 
     @verbatim ( language="comment", text="Documentation of MyMessage." "Adjacent string literal." )
+    @transfer_mode(SHMEM_REF)
     struct MyMessage {
       short short_value, short_value2;
       @default ( value=123 )

--- a/rosidl_parser/test/test_parser.py
+++ b/rosidl_parser/test/test_parser.py
@@ -160,7 +160,7 @@ def test_message_parser_annotations(message_idl_file):
     assert 'text' in structure.annotations[0].value
     assert structure.annotations[0].value['text'] == \
         'Documentation of MyMessage.Adjacent string literal.'
-        
+
     assert structure.annotations[1].name == 'transfer_mode'
     assert structure.annotations[1].value == 'SHMEM_REF'
 

--- a/rosidl_parser/test/test_parser.py
+++ b/rosidl_parser/test/test_parser.py
@@ -152,7 +152,7 @@ def test_message_parser_annotations(message_idl_file):
     assert len(messages) == 1
     structure = messages[0].structure
 
-    assert len(structure.annotations) == 1
+    assert len(structure.annotations) == 2
     assert structure.annotations[0].name == 'verbatim'
     assert len(structure.annotations[0].value) == 2
     assert 'language' in structure.annotations[0].value
@@ -160,6 +160,9 @@ def test_message_parser_annotations(message_idl_file):
     assert 'text' in structure.annotations[0].value
     assert structure.annotations[0].value['text'] == \
         'Documentation of MyMessage.Adjacent string literal.'
+        
+    assert structure.annotations[1].name == 'transfer_mode'
+    assert structure.annotations[1].value == 'SHMEM_REF'
 
     assert len(structure.members[2].annotations) == 1
 

